File: Sample-java/src/main/java/com/opera/ads/demo/fragment/NativeFragment.java
package com.opera.ads.demo.fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.opera.ads.AdError;
import com.opera.ads.AdFormat;
import com.opera.ads.demo.R;
import com.opera.ads.demo.databinding.NativeAdLayoutBinding;
import com.opera.ads.demo.util.Constant;
import com.opera.ads.nativead.NativeAd;
import com.opera.ads.nativead.NativeAdListener;
import com.opera.ads.nativead.NativeAdLoader;

public class NativeFragment extends BaseFragment {
    @Nullable
    private NativeAd mNativeAd;

    @NonNull
    @Override
    protected AdFormat getAdFormat() {
        return AdFormat.NATIVE;
    }

    @Override
    protected int getAdFormatStringId() {
        return R.string.native_ad_format;
    }

    @Override
    protected void loadAd() {
        if (mLogView != null) mLogView.print("Loading...");
        String placementId = mPlacementId == null ? Constant.getPlacementId(AdFormat.NATIVE, false) : mPlacementId;
        NativeAdLoader.loadAd(requireContext(), placementId, new NativeAdListener()  {
            @Override
            public void onAdLoaded(@NonNull NativeAd nativeAd) {
                mNativeAd = nativeAd;
                enableShowAd();
                enableDestroyAd();
                if (mLogView != null) {
                    mLogView.print("Loaded, ad: " + mNativeAd.title());
                }
            }

            @Override
            public void onAdFailedToLoad(@NonNull AdError error) {
                if (mLogView != null) {
                    mLogView.print("Failed, error: " + error.getMessage());
                }
            }

            @Override
            public void onAdImpression() {
                if (mLogView != null) {
                    mLogView.print("Impression, ad: " + mNativeAd.title());
                }
            }

            @Override
            public void onAdClicked() {
                if (mLogView != null) {
                    mLogView.print("Clicked, ad: " + mNativeAd.title());
                }
            }
        });
    }

    @Override
    protected void showAd() {
        if (mNativeAd != null && mAdContainer != null) {
            if (mNativeAd.isAdInvalidated()) {
                mLogView.print("Ad is invalidated.");
                destroyAd();
                return;
            }
            mNativeAd.setAdChoicePosition(NativeAd.AdChoicePosition.TOP_RIGHT);
            NativeAdLayoutBinding itemBinding = NativeAdLayoutBinding.inflate(LayoutInflater.from(requireContext()));
            itemBinding.nativeAdTitle.setText(mNativeAd.title());
            itemBinding.nativeAdBody.setText(mNativeAd.description());
            final Double starRating = mNativeAd.starRating();
            itemBinding.starRating.setText(starRating != null ? "Rating: " + starRating : "");
            itemBinding.starRating.setVisibility(starRating != null ? View.VISIBLE : View.GONE);
            itemBinding.nativeAdCallToAction.setText(mNativeAd.callToAction());
            itemBinding.nativeAdMedia.setImageScaleType(ImageView.ScaleType.CENTER_CROP);

            NativeAd.InteractionViews interactionViews = new NativeAd.InteractionViews.Builder(
                    itemBinding.nativeAdMedia
            ).setTitleView(itemBinding.nativeAdTitle)
                    .setBodyView(itemBinding.nativeAdBody)
                    .setCallToActionView(itemBinding.nativeAdCallToAction)
                    .setIconView(itemBinding.nativeAdIcon)
                    .build();

            FrameLayout nativeAdRootView = itemBinding.getRoot();
            mNativeAd.registerInteractionViews(nativeAdRootView, interactionViews);
            if (mAdContainer != null) mAdContainer.addView(nativeAdRootView);
            disableShowAd();
            enableDestroyAd();
        }
    }

    @Override
    protected void destroyAd() {
        if (mNativeAd != null) {
            mNativeAd.destroy();
            if (mAdContainer != null) mAdContainer.removeAllViews();
            if (mLogView != null) mLogView.print("Destroyed...");
            mNativeAd = null;
            disableShowAd();
            disableDestroyAd();
        }
    }
}

File: Sample-kotlin/src/main/kotlin/com/opera/ads/demo/fragment/NativeFragment.kt
package com.opera.ads.demo.fragment

import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import com.opera.ads.AdError
import com.opera.ads.AdFormat
import com.opera.ads.demo.R
import com.opera.ads.demo.databinding.NativeAdLayoutBinding
import com.opera.ads.nativead.NativeAd
import com.opera.ads.nativead.NativeAdListener
import com.opera.ads.nativead.NativeAdLoader

class NativeFragment : BaseFragment(AdFormat.NATIVE, false, false, R.string.native_ad_format) {
    private var nativeAd: NativeAd? = null

    override fun loadAd() {
        logger.print("Loading...")
        NativeAdLoader.loadAd(
            requireContext(),
            placementId!!,
            object : NativeAdListener {
                override fun onAdLoaded(ad: NativeAd) {
                    nativeAd = ad
                    enableShowAd()
                    enableDestroyAd()
                    logger.print("Loaded, ad: ${nativeAd!!.title()}")
                }

                override fun onAdFailedToLoad(error: AdError) {
                    logger.print("Failed, error: ${error.message}")
                }

                override fun onAdImpression() {
                    logger.print("Impression, ad: ${nativeAd!!.title()}")
                }

                override fun onAdClicked() {
                    logger.print("Clicked, ad: ${nativeAd!!.title()}")
                }
            }
        )
    }

    override fun showAd() {
        nativeAd?.let {
            if (it.isAdInvalidated()) {
                logger.print("Ad is invalidated.")
                destroyAd()
                return
            }
            it.setAdChoicePosition(NativeAd.AdChoicePosition.TOP_RIGHT)
            NativeAdLayoutBinding.inflate(LayoutInflater.from(requireContext())).apply {
                nativeAdTitle.text = it.title()
                nativeAdBody.text = it.description()
                it.starRating()?.let { rating ->
                    starRating.text = "Rating: $rating"
                    starRating.visibility = View.VISIBLE
                } ?: run {
                    starRating.visibility = View.GONE
                }
                nativeAdCallToAction.text = it.callToAction()
                nativeAdMedia.setImageScaleType(ImageView.ScaleType.CENTER_CROP)

                val nativeAdRootView: FrameLayout = getRoot()
                adContainer().addView(nativeAdRootView)
                val interactionViews = NativeAd.InteractionViews.Builder(nativeAdMedia)
                    .setTitleView(nativeAdTitle)
                    .setBodyView(nativeAdBody)
                    .setCallToActionView(nativeAdCallToAction)
                    .setIconView(nativeAdIcon)
                    .build()
                it.registerInteractionViews(nativeAdRootView, interactionViews)
            }
            disableShowAd()
            enableDestroyAd()
        }
    }

    override fun destroyAd() {
        nativeAd?.destroy()
        nativeAd = null
        adContainer().removeAllViews()
        logger.print("Destroyed...")
        disableShowAd()
        disableDestroyAd()
    }
}
